### ISSUE TITLE
Always delete cookie from a domain, or never save cookie (Blacklist)

### ISSUE THREAD

**Issue by zmulk** [Cookie-AutoDelete/Cookie-AutoDelete#134]
### Describe the Feature
I want to add a domain or website where the cookies never will be saved, or will be deleted directly.

- OS/version:  Windows 10
- Browser/version:  Vivaldi
- Cookie AutoDelete version: 1.4.3

**Comment by pwd-github** [comment]
That options exists with the "Clean After Tab Close" option.

**Comment by mrdokenny** [comment]
So a blacklist basically?

**Comment by MrGlasspoole** [comment]
You closed my Feature Request #210 as duplicate of this one.
This one talks about deleting cookies from sites on a blacklist.
I talk about rejecting all cookies by default (not first allowing and then deleting them).

Also at #156 i can read "134 resolved" but i don't see a blacklist?

**Comment by ke-d** [comment]
@MrGlasspoole 
>This one talks about deleting cookies from sites on a blacklist.
I talk about rejecting all cookies by default (not first allowing and then deleting them).

It is sort of related since if I ever implement a blacklist, it would be similar to a greylist where you can add a wildcard in the list to make that the default.

Also, the only thing I can do with current APIs is basically checking the [onChanged event](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/cookies/onChanged) after a website has already set cookies and remove them after, which is not really a true "blacklist".

Another thing with #44 is that it requires at least 1 cookie to be set so that the browsingData API knows which domains to clear localstorage from. That part would not work if people start using a blacklist.

Anyways I'm open to suggestions on this.

>Also at #156 i can read "134 resolved" but i don't see a blacklist?

That was a mistake. I selected the wrong issue but forgot to change it until you mentioned it.

**Comment by ElhemEnohpi** [comment]
@zmulk you can use the built-in blacklist feature of the browser. In Vivaldi go to `vivaldi://settings/content/cookies` and add a "block" exception for the site. This will prevent cookies and local storage from being used.

The Cookie AutoDelete description says "You must enable cookies in the browser as this extension will replace the internal cookie manager", but that's not really accurate. The internal cookie manager isn't replaced, it still works as usual. You can use it to block certain sites' cookies, or even to block all cookies by default and only allow certain ones, which will then be auto-deleted. (Or not, if you also whitelist them in C-AD).

**Comment by drcheap** [comment]
> Also, the only thing I can do with current APIs is basically checking the onChanged event after a website has already set cookies and remove them after, which is not really a true "blacklist".

Given they would only last for a matter of milliseconds, or at least not long enough to ever be sent to a server in a future request, that would be equally as effective as a true blacklist.

**Comment by thomasrussellmurphy** [comment]
I would also be interested in seeing a feature to immediately delete cookies from select domains. Setting the blacklist to `*` (processed after whitelist/graylist) would be one option, but I'd welcome the option to only instantly remove cookies set by unwanted sites.

**Comment by grenzor** [comment]
Looks like functionality to do this landed in Nightly: https://bugzilla.mozilla.org/show_bug.cgi?id=1363860

Btw, is this a dupe of #95?

**Comment by ElhemEnohpi** [comment]
@grenzor , no, that doesn't help with this issue. And #95 is not the same issue as this one, though they both could maybe be fixed with the same solution, depending how it's implemented.

**Comment by kathampy** [comment]
I'd also like to blacklist a domain in the default container alone, but not in the separate container I use for that domain. This prevents me from using the browser settings to block cookies for the domain.

**Comment by Existencielle** [comment]
Hello,
my original request #450 ("Add "Blacklist" option beside Greylist and Whitelist") was closed, saying it was a duplicate of this one.

Here I do find the suggestion that I should just block cookies for that domain in Firefox settings.
I tried that, but it does not have the desired effect. Blocking cookies completely sometimes prevents pages from loading properly. What I want is the cookies to be accepted, but then after the grace period (I have set 6 seconds, don't remember if that is the default) them to be thrown away. 

It may still be appropiate to merge my issue into this one, however I wanted to point out that the workaround suggested in this thread is not helping.

**Comment by pavinjosdev** [comment]
@zmulk Cookies can be blocked for specific sites by the browser's own cookie manager/settings. [Docs ](https://support.google.com/chrome/answer/95647)for Chrome.

> Blocking cookies completely sometimes prevents pages from loading properly.

@Existencielle Good point, I haven't found an open source extension to delete cookies every `x` seconds but as a workaround what I do is set the cookie to `read only` or `protected` status from an extension like [EditThisCookie](https://github.com/ETCExtensions/Edit-This-Cookie)

**Comment by ElhemEnohpi** [comment]
Fyi, [Forget Me Not](https://addons.mozilla.org/en-US/firefox/addon/forget_me_not/) supports blacklist rules, as well as a default "block" action for sites that don't have any rule set for them.

**Comment by ghost** [comment]
[CookieMaster](https://addons.mozilla.org/en-US/firefox/addon/cookiemaster/)  

> allows you to control which sites are allowed to set cookies

**Comment by dhowe** [comment]
+1 for a true blacklist = don't allow cookie to be set (you can do this with the current API)

**Comment by Existencielle** [comment]
> 
> 
> +1 for a true blacklist = don't allow cookie to be set (you can do this with the current API)

You do realize you don't need an addon for that, you can just set that in the Options?

I think the features that Cookie-AutoDelete provides should be stuff you cannot do without it.

**Comment by kathampy** [comment]
> You do realize you don't need an addon for that, you can just set that in the Options?
> I think the features that Cookie-AutoDelete provides should be stuff you cannot do without it.

You cannot block cookies per container. I want to block a certain domain's cookies in the Default container, but allow them in the domain's dedicated container.

**Comment by kennethtran93** [comment]
Apologies for temporarily closing this feature - I've adjusted the main issue to reflect the feature.

**Comment by thomasrussellmurphy** [comment]
Thank you for re-opening and cleaning this up. Looking forward to this potential enhancement.

**Comment by MPeti1** [comment]
So far Forget Me Not seemed to be the best for the task. Though from C-DA's name, I would have thought it can do proper blacklisting too.. (not cookie blocking, but deletion on tab close or whatever, so inverting the current functionality). At least now that I know that the dev doesn't want that, I won't check it every few months if they've added that feature

**Comment by kennethtran93** [comment]
This requires an overhaul of the UI elements within the popup as well as how it is displayed in the List of Expressions Table and will take some time.

**Comment by CaseOf** [comment]
> You closed my Feature Request #210 as duplicate of this one. This one talks about deleting cookies from sites on a blacklist. I talk about rejecting all cookies by default (not first allowing and then deleting them).
> 
> Also at #156 i can read "134 resolved" but i don't see a blacklist?

Hello,
I am actually using CookieBro simultaneously with Cookie AutoDelete for a blacklist mode by default.
Adding `*.*` in blacklist in CookieBro makes it behave in blacklist mode by default.
Maybe could it be inspired for this mode in Cookie AutoDelete?
It matches well what I would like to see in Cookie AutoDelete for blacklist.
What is missing in CookieBro though is that blacklist mode to be set also on localstorage and indexeddb
Edit: But I would try ForgetMeNot and CookieMaster cited above.

**Comment by kilobyte** [comment]
Firefox itself already implements a blacklist, its UI is just very unwieldy.  You need to go: Edit → Settings → Privacy & Security → Cookies and Site Data/Manage Exceptions... → enter the domain name, click "Block", "Save changes".

It would be great if you could add a button next to [ + Greylist ][ +Whitelist ] that says [ + Blacklist ] and adds to the browser's (rather than extension's) blacklist.

**Comment by kennethtran93** [comment]
> Firefox itself already implements a blacklist, its UI is just very unwieldy. You need to go: Edit → Settings → Privacy & Security → Cookies and Site Data/Manage Exceptions... → enter the domain name, click "Block", "Save changes".
> 
> It would be great if you could add a button next to [ + Greylist ][ +Whitelist ] that says [ + Blacklist ] and adds to the browser's (rather than extension's) blacklist.

The Firefox blocklist is per entire sites, not per cookie on a site.  Some users may want to save a specific cookie and/or only block certain cookies on a site.

Unfortunately, the Web Extension API has no way (at least in Firefox) to read/modify the native browser's blocklist.

**Comment by curiosityseeker** [comment]
> Firefox itself already implements a blacklist, its UI is just very unwieldy. You need to go: Edit → Settings → Privacy & Security → Cookies and Site Data/Manage Exceptions... → enter the domain name, click "Block", "Save changes".


There is much faster way to accomplish that: Ctrl + i -> _Permissions_

**Comment by jerome-tgl** [comment]
>It would be great if you could add a button next to [ + Greylist ][ +Whitelist ] that says [ + Blacklist ] and adds to the browser's (rather than extension's) blacklist.

Yes. Here's a use case to illustrate  I need CAD to clean youtube.com cookies regularly to prevent Youtube from showing recommendations based on my history. But I usually one or a few tab left open so CAD won't auto-delete until I close them all and open none for 15 seconds (default delay). Opening one Youtube URL while there's no tab left will prevent CAD from cleaning this history. So I have to clean up all tabs to ensure CAD does its job.

A blacklist feature would allow me to escape Youtube's filter bubble at once.
